packages/components/src/Dialog/Dialog.tsx

```diff
@@ -9,16 +9,73 @@ import type RAC from 'react-aria-components';
 import { Dialog, OverlayTriggerStateContext } from 'react-aria-components';
 import { cn, useClassNames } from '@marigold/system';
 import { CloseButton } from '../CloseButton';
-import { Modal } from '../Overlay';
+import { Modal, ModalProps } from '../Overlay';
+import { DialogContext } from './Context';
 import { DialogActions } from './DialogActions';
 import { DialogContent } from './DialogContent';
 import { DialogTitle } from './DialogTitle';
-import { DialogContext, DialogTrigger } from './DialogTrigger';
+import { DialogTrigger } from './DialogTrigger';
 
-export interface DialogProps
+// Helper
+// ---------------
+interface InnerDialogProps
   extends Omit<RAC.DialogProps, 'className' | 'style'> {
+  className: {
+    container: string;
+    closeButton: string;
+  };
+  closeButton?: boolean;
+  children?:
+    | React.ReactNode
+    | ((args: { close: () => void }) => React.ReactNode);
+}
+
+/**
+ * Needed so that the close button and function can be used inside the dialog,
+ * when the dialog is controlled and no <Dialog.Trigger> is used.
+ */
+const InnerDialog = forwardRef(
+  (
+    { className, closeButton, ...props }: InnerDialogProps,
+    ref: Ref<HTMLElement> | undefined
+  ) => {
+    const state = useContext(OverlayTriggerStateContext);
+    const children =
+      typeof props.children === 'function'
+        ? props.children({
+            close: state?.close ?? (() => {}),
+          })
+        : props.children;
+
+    return (
+      <Dialog
+        {...props}
+        ref={ref}
+        className={cn(
+          'relative mx-auto outline-hidden',
+          "grid [grid-template-areas:'title'_'content'_'actions']",
+          className.container
```

Comment:
why not get this from the `useClassNames` hook?

Reply:
moved to inner dialog. good catch!